### PR TITLE
audit: check for Python-wide site-package usage

### PR DESCRIPTION
**Pull request by iMichka**
See https://github.com/Homebrew/brew/pull/16662

We would like to enforce vendoring for Python libraries, or the usage of a virtualenv in the formula's libexec directory, using a virtualenv.

- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

**Comment by iMichka** [comment]
This needs an allowlist and further testing. It's just an early draft to see if this audit would make sense to enforce our vendoring policy.

Note: I am aware that the existing check_easy_install_pth check looks fishy nowadays, and will handle that in a separate pull request.

**Comment by Bo98** [comment]
Seems like it would catch a ton of things that happen to ship Python bindings which we have always allowed, unless we're planning on changing that?

Limiting this to `pythonhosted.org` URLs would however could make sense. I'd probably limit this to Homebrew/core only too as this doesn't make sense to restrict for third-party taps.

**Comment by MikeMcQuaid** [comment]
> Python bindings which we have always allowed, unless we're planning on changing that?

Don't think we should change that 👍🏻 

Could maybe do something like always allow that if there's also something anything `lib`?

**Comment by iMichka** [comment]
> Limiting this to `pythonhosted.org` URLs would however could make sense. I'd probably limit this to Homebrew/core only too as this doesn't make sense to restrict for third-party taps.

Done

**Comment by iMichka** [comment]
> > Python bindings which we have always allowed, unless we're planning on changing that?
> 
> Don't think we should change that 👍🏻
> 
> Could maybe do something like always allow that if there's also something anything `lib`?

I made an attempt to parse for .so files (will need to adapt for .dylib on Linux too).
This makes numpy pass, as it makes also python-cryptography pass (it contains a .so file). Basically this means we allow any .so/.dylib file we compiled from source.

The only remaining exception I am aware of is `certifi`, that is not correctly handled by my audit.
The check needs some testing on more formulae to be sure we cover all the cases (tests would be good too but might be hard to implement).

**Comment by iMichka** [comment]
I opened another (unrelated) PR for the check_easy_install_pth check:
https://github.com/Homebrew/brew/pull/16698

File: Library/Homebrew/formula_cellar_checks.rb [hunk 1]
```diff
@@ -143,6 +143,25 @@
     EOS
   end
 
+  def check_global_site_package_usage(formula)
+    return if !formula.tap.core_tap?
+    return if !formula.stable.url.match?("https://files.pythonhosted.org")
```
[the remark this answers is not in the document]
Comment (dduugg) [reply Library/Homebrew/formula_cellar_checks.rb:148]: Replacing `match?` with `include?` or `start_with?` should do the trick here.

**Comment by iMichka** [comment]
Made some tests with the current implementation.

These are ok:
```
numpy -> Does not fail audit, due to Python bindings
scipy ->  Does not fail audit, due to Python bindings
python-matplotlib ->  Does not fail audit, due to Python bindings
python-cryptography ->  Does not fail audit, due to Python bindings
six -> Fails audit
pygments -> Fails audit
python-packaging -> Fails audit
python-pytz -> Fails audit
python-urllib3 -> Fails audit
python-tabulate -> Fails audit
```

These are not:
```
python-certifi -> Fails audit, but we want to make it pass
poetry ->  Does not fail audit, but should?
python-markupsafe ->  Does not fail audit (but should fail? Contains _speedups.cpython-312-darwin.so?)
```

**Comment by Bo98** [comment]
> poetry ->  Does not fail audit, but should?

I don't think so - it's a CLI tool. The underlying backend is `poetry-core`.

**Pull request closed without merging**